File: src/sim/relational-time.ts
/**
 * Two-clock / relational time.
 *
 * Page-Wootters time is not absolute: it is recovered by conditioning on a
 * chosen clock subsystem. Different clocks define different emergent time
 * coordinates for the same underlying quantum history.
 *
 * We evolve one physical trajectory and read it off with two clocks:
 *
 *   Clock A (uniform): fixed Δt ticks — a "global" discretization.
 *   Clock B (physical): ticks when the disturbance signal at a chosen site
 *     crosses evenly spaced thresholds — a clock *inside* the system that
 *     stalls until the quench reaches it, then runs fast.
 *
 * The map t_A ↔ t_B is generally nonlinear: neither clock is "the" time.
 */

import {
  cloneState,
  expectationZ,
  Hamiltonian,
  normalize,
  type QuantumState,
} from './quantum.ts'
import { analyzeEmergentGeometry } from './geometry.ts'
import {
  evolveInterval,
  type SpacetimeResult,
  type SpacetimeSlice,
} from './spacetime.ts'

export interface TrajectoryPoint {
  t: number
  psi: QuantumState
  ref: QuantumState | null
}

export interface DualClockConfig {
  hamiltonian: Hamiltonian
  initial: QuantumState
  reference?: QuantumState
  /** Physical time step for the stored trajectory. */
  dt: number
  /** Number of trajectory points (uniform clock uses all of them). */
  steps: number
  /** Qubit whose local signal defines the physical clock. */
  clockSite: number
  /** Number of physical-clock slices to extract. */
  physicalSlices: number
  embedDim?: number
  order?: number
}

export interface ClockHistory {
  label: string
  /** Which qubit defines this clock, if any. */
  clockSite?: number
  result: SpacetimeResult
}

export interface DualClockResult {
  sites: number
  /** Uniform Δt clock. */
  uniform: ClockHistory
  /** Event-based clock at `clockSite`. */
  physical: ClockHistory
  /**
   * Relational time map: at physical-clock tick k, what uniform-clock reading
   * index records the same system state? Nonlinearity means the clocks disagree.
   */
  timeMap: { tauUniform: number; tauPhysical: number }[]
  /** R² of linear fit tauUniform = a * tauPhysical + b (1 = synchronized). */
  syncR2: number
  /** Slope of affine fit tauUniform = slope * tauPhysical + intercept. */
  syncSlope: number
  syncIntercept: number
  elapsedMs: number
}

function buildSlice(
  hamiltonian: Hamiltonian,
  psi: QuantumState,
  ref: QuantumState | null,
  baseZ: number[],
  t: number,
  k: number,
  embedDim: number,
): SpacetimeSlice {
  const sites = hamiltonian.n
  const energy = hamiltonian.expectation(psi)
  const report = analyzeEmergentGeometry(psi, 1, Math.max(embedDim, 1))
  let coords: number[][]
  if (embedDim >= 2) {
    coords = report.mds.coords.map((c) => [c[0] ?? 0, c[1] ?? 0])
  } else {
    const raw = report.mds.coords.map((c) => c[0] ?? 0)
    let cov = 0
    const meanI = (sites - 1) / 2
    const meanC = raw.reduce((a, b) => a + b, 0) / sites
    for (let i = 0; i < sites; i++) cov += (i - meanI) * (raw[i] - meanC)
    const oriented = cov < 0 ? raw.map((c) => -c) : raw
    coords = oriented.map((x) => [x])
  }
  const zExpectation = Array.from({ length: sites }, (_, q) => expectationZ(psi, q))
  const refZ = ref
    ? Array.from({ length: sites }, (_, q) => expectationZ(ref, q))
    : baseZ
  const signal = zExpectation.map((z, q) => Math.abs(z - refZ[q]))
  return { k, t, coords, zExpectation, signal, energy }
}

/** Evolve and store the full trajectory at fixed Δt. */
export function evolveTrajectory(config: {
  hamiltonian: Hamiltonian
  initial: QuantumState
  reference?: QuantumState
  dt: number
  steps: number
  order?: number
}): TrajectoryPoint[] {
  const { hamiltonian, initial, reference, dt, steps, order = 6 } = config
  const radius = Math.max(hamiltonian.estimateSpectralRadius(() => 0.5), 1e-6)
  const points: TrajectoryPoint[] = []
  let psi = cloneState(initial)
  let ref = reference ? cloneState(reference) : null
  for (let k = 0; k < steps; k++) {
    points.push({ t: k * dt, psi: cloneState(psi), ref: ref ? cloneState(ref) : null })
    psi = evolveInterval(hamiltonian, psi, dt, radius, order)
    normalize(psi)
    if (ref) {
      ref = evolveInterval(hamiltonian, ref, dt, radius, order)
      normalize(ref)
    }
  }
  return points
}

function signalAtSite(point: TrajectoryPoint, site: number, baseZ: number[]): number {
  const z = expectationZ(point.psi, site)
  const refZ = point.ref ? expectationZ(point.ref, site) : baseZ[site]
  return Math.abs(z - refZ)
}

function slicesFromIndices(
  hamiltonian: Hamiltonian,
  trajectory: TrajectoryPoint[],
  indices: number[],
  times: number[],
  initial: QuantumState,
  embedDim: number,
): SpacetimeSlice[] {
  const baseZ = Array.from({ length: hamiltonian.n }, (_, q) =>
    expectationZ(initial, q),
  )
  return indices.map((idx, k) =>
    buildSlice(
      hamiltonian,
      trajectory[idx].psi,
      trajectory[idx].ref,
      baseZ,
      times[k],
      k,
      embedDim,
    ),
  )
}

/** Pick physical-clock ticks when the signal at `clockSite` crosses thresholds. */
function physicalClockIndices(
  trajectory: TrajectoryPoint[],
  clockSite: number,
  baseZ: number[],
  numSlices: number,
): { indices: number[]; times: number[] } {
  const signals = trajectory.map((p) => signalAtSite(p, clockSite, baseZ))
  const maxSig = Math.max(...signals, 1e-9)
  const targets = Array.from({ length: numSlices }, (_, i) =>
    (maxSig * (i + 0.5)) / numSlices,
  )

  const indices: number[] = [0]
  const times: number[] = [0]
  let searchFrom = 1
  for (const target of targets) {
    let found = -1
    for (let i = searchFrom; i < trajectory.length; i++) {
      if (signals[i] >= target) {
        found = i
        break
      }
    }
    if (found < 0) break
    indices.push(found)
    times.push(trajectory[found].t)
    searchFrom = found + 1
  }

  // Ensure at least two slices for plotting.
  if (indices.length < 2 && trajectory.length > 1) {
    indices.push(trajectory.length - 1)
    times.push(trajectory[trajectory.length - 1].t)
  }
  return { indices, times }
}

function fitAffine(xs: number[], ys: number[]): { slope: number; intercept: number; r2: number } {
  if (xs.length < 2) return { slope: 1, intercept: 0, r2: 1 }
  const n = xs.length
  let sxx = 0
  let sxy = 0
  let sx = 0
  let sy = 0
  for (let i = 0; i < n; i++) {
    sxx += xs[i] * xs[i]
    sxy += xs[i] * ys[i]
    sx += xs[i]
    sy += ys[i]
  }
  const denom = n * sxx - sx * sx
  const slope = denom !== 0 ? (n * sxy - sx * sy) / denom : 1
  const intercept = (sy - slope * sx) / n
  const yBar = sy / n
  let ssRes = 0
  let ssTot = 0
  for (let i = 0; i < n; i++) {
    const pred = slope * xs[i] + intercept
    ssRes += (ys[i] - pred) ** 2
    ssTot += (ys[i] - yBar) ** 2
  }
  const r2 = ssTot > 0 ? 1 - ssRes / ssTot : 1
  return { slope, intercept, r2 }
}

/** Build two emergent histories from one shared quantum trajectory. */
export function buildDualClock(config: DualClockConfig): DualClockResult {
  const start = performance.now()
  const {
    hamiltonian,
    initial,
    reference,
    dt,
    steps,
    clockSite,
    physicalSlices,
    embedDim = 1,
    order = 6,
  } = config
  const sites = hamiltonian.n

  const trajectory = evolveTrajectory({
    hamiltonian,
    initial,
    reference,
    dt,
    steps,
    order,
  })

  const baseZ = Array.from({ length: sites }, (_, q) => expectationZ(initial, q))
  const energy0 = hamiltonian.expectation(initial)

  // Uniform clock: every trajectory point, t = k * dt.
  const uniformIndices = trajectory.map((_, k) => k)
  const uniformTimes = trajectory.map((p) => p.t)
  const uniformSliceList = slicesFromIndices(
    hamiltonian,
    trajectory,
    uniformIndices,
    uniformTimes,
    initial,
    embedDim,
  )
  let energyDrift = 0
  for (const s of uniformSliceList) {
    energyDrift = Math.max(energyDrift, Math.abs(s.energy - energy0))
  }

  // Physical clock: event-based at clockSite.
  const { indices: physIndices } = physicalClockIndices(
    trajectory,
    clockSite,
    baseZ,
    physicalSlices,
  )
  const physicalSliceList = slicesFromIndices(
    hamiltonian,
    trajectory,
    physIndices,
    // Physical clock's own relational time coordinate = tick index, not Schrödinger t.
    physIndices.map((_, k) => k),
    initial,
    embedDim,
  )
  for (const s of physicalSliceList) {
    energyDrift = Math.max(energyDrift, Math.abs(s.energy - energy0))
  }

  // Relational time map: uniform reading index vs physical tick index at same state.
  const timeMap = physIndices.map((uniformIdx, kPhysical) => ({
    tauUniform: uniformIdx,
    tauPhysical: kPhysical,
  }))

  const { slope: syncSlope, intercept: syncIntercept, r2: syncR2 } = fitAffine(
    timeMap.map((p) => p.tauPhysical),
    timeMap.map((p) => p.tauUniform),
  )

  const elapsedMs = performance.now() - start

  return {
    sites,
    uniform: {
      label: 'Uniform clock (fixed Δt)',
      result: {
        sites,
        slices: uniformSliceList,
        energyDrift,
        elapsedMs,
      },
    },
    physical: {
      label: `Physical clock (site ${clockSite})`,
      clockSite,
      result: {
        sites,
        slices: physicalSliceList,
        energyDrift,
        elapsedMs,
      },
    },
    timeMap,
    syncR2,
    syncSlope,
    syncIntercept,
    elapsedMs,
  }
}

File: wasm/mad-dog-sim/src/relational_time.rs
//! Two-clock / relational time — port of `src/sim/relational-time.ts`.

use crate::geometry::analyze_emergent_geometry;
use crate::quantum::{evolve_interval, expectation_z, Hamiltonian, QuantumState};
use crate::rng::Rng;
use crate::spacetime::{SpacetimeResult, SpacetimeSlice};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug)]
pub struct TrajectoryPoint {
    pub t: f64,
    pub psi: QuantumState,
    pub reference: Option<QuantumState>,
}

#[derive(Clone, Debug)]
pub struct DualClockConfig<'a> {
    pub hamiltonian: &'a Hamiltonian,
    pub initial: QuantumState,
    pub reference: Option<QuantumState>,
    pub dt: f64,
    pub steps: usize,
    pub clock_site: usize,
    pub physical_slices: usize,
    pub embed_dim: usize,
    pub order: usize,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClockHistory {
    pub label: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub clock_site: Option<usize>,
    pub result: SpacetimeResult,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimeMapPoint {
    pub tau_uniform: f64,
    pub tau_physical: f64,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DualClockResult {
    pub sites: usize,
    pub uniform: ClockHistory,
    pub physical: ClockHistory,
    pub time_map: Vec<TimeMapPoint>,
    pub sync_r2: f64,
    pub sync_slope: f64,
    pub sync_intercept: f64,
    pub elapsed_ms: f64,
}

fn orient_coords(coords: &[f64]) -> Vec<f64> {
    let n = coords.len();
    if n < 2 {
        return coords.to_vec();
    }
    let mean_i = (n - 1) as f64 / 2.0;
    let mean_c = coords.iter().sum::<f64>() / n as f64;
    let mut cov = 0.0;
    for (i, &c) in coords.iter().enumerate() {
        cov += (i as f64 - mean_i) * (c - mean_c);
    }
    if cov < 0.0 {
        coords.iter().map(|c| -c).collect()
    } else {
        coords.to_vec()
    }
}

fn build_slice(
    hamiltonian: &Hamiltonian,
    psi: &QuantumState,
    reference: Option<&QuantumState>,
    base_z: &[f64],
    t: f64,
    k: usize,
    embed_dim: usize,
) -> SpacetimeSlice {
    let sites = hamiltonian.n;
    let energy = hamiltonian.expectation(psi);
    let report = analyze_emergent_geometry(psi, 1.0, embed_dim.max(1));
    let coords = if embed_dim >= 2 {
        report
            .mds
            .coords
            .iter()
            .map(|c| vec![c.first().copied().unwrap_or(0.0), c.get(1).copied().unwrap_or(0.0)])
            .collect()
    } else {
        let raw: Vec<f64> = report
            .mds
            .coords
            .iter()
            .map(|c| c.first().copied().unwrap_or(0.0))
            .collect();
        orient_coords(&raw).into_iter().map(|x| vec![x]).collect()
    };
    let z_expectation: Vec<f64> = (0..sites).map(|q| expectation_z(psi, q)).collect();
    let ref_z: Vec<f64> = if let Some(r) = reference {
        (0..sites).map(|q| expectation_z(r, q)).collect()
    } else {
        base_z.to_vec()
    };
    let signal: Vec<f64> = z_expectation
        .iter()
        .zip(ref_z.iter())
        .map(|(z, rz)| (z - rz).abs())
        .collect();
    SpacetimeSlice {
        k,
        t,
        coords,
        z_expectation,
        signal,
        energy,
    }
}

pub fn evolve_trajectory(
    hamiltonian: &Hamiltonian,
    initial: &QuantumState,
    reference: Option<&QuantumState>,
    dt: f64,
    steps: usize,
    order: usize,
) -> Vec<TrajectoryPoint> {
    let mut rng = Rng::new(42);
    let radius = hamiltonian
        .estimate_spectral_radius(&mut rng, 30)
        .max(1e-6);
    let mut points = Vec::with_capacity(steps);
    let mut psi = initial.clone_state();
    let mut ref_state = reference.cloned();
    for k in 0..steps {
        points.push(TrajectoryPoint {
            t: k as f64 * dt,
            psi: psi.clone_state(),
            reference: ref_state.as_ref().map(|r| r.clone_state()),
        });
        psi = evolve_interval(hamiltonian, &psi, dt, radius, order);
        psi.normalize();
        if let Some(ref mut r) = ref_state {
            *r = evolve_interval(hamiltonian, r, dt, radius, order);
            r.normalize();
        }
    }
    points
}

fn signal_at_site(point: &TrajectoryPoint, site: usize, base_z: &[f64]) -> f64 {
    let z = expectation_z(&point.psi, site);
    let ref_z = point
        .reference
        .as_ref()
        .map(|r| expectation_z(r, site))
        .unwrap_or(base_z[site]);
    (z - ref_z).abs()
}

fn slices_from_indices(
    hamiltonian: &Hamiltonian,
    trajectory: &[TrajectoryPoint],
    indices: &[usize],
    times: &[f64],
    initial: &QuantumState,
    embed_dim: usize,
) -> Vec<SpacetimeSlice> {
    let base_z: Vec<f64> = (0..hamiltonian.n)
        .map(|q| expectation_z(initial, q))
        .collect();
    indices
        .iter()
        .zip(times.iter())
        .enumerate()
        .map(|(k, (&idx, &t))| {
            let point = &trajectory[idx];
            build_slice(
                hamiltonian,
                &point.psi,
                point.reference.as_ref(),
                &base_z,
                t,
                k,
                embed_dim,
            )
        })
        .collect()
}

fn physical_clock_indices(
    trajectory: &[TrajectoryPoint],
    clock_site: usize,
    base_z: &[f64],
    num_slices: usize,
) -> (Vec<usize>, Vec<f64>) {
    let signals: Vec<f64> = trajectory
        .iter()
        .map(|p| signal_at_site(p, clock_site, base_z))
        .collect();
    let max_sig = signals.iter().copied().fold(1e-9_f64, f64::max);
    let targets: Vec<f64> = (0..num_slices)
        .map(|i| max_sig * (i as f64 + 0.5) / num_slices as f64)
        .collect();

    let mut indices = vec![0];
    let mut times = vec![0.0];
    let mut search_from = 1;
    for target in targets {
        let mut found = None;
        for i in search_from..trajectory.len() {
            if signals[i] >= target {
                found = Some(i);
                break;
            }
        }
        let Some(i) = found else { break };
        indices.push(i);
        times.push(trajectory[i].t);
        search_from = i + 1;
    }

    if indices.len() < 2 && trajectory.len() > 1 {
        let last = trajectory.len() - 1;
        indices.push(last);
        times.push(trajectory[last].t);
    }
    (indices, times)
}

pub fn fit_affine(xs: &[f64], ys: &[f64]) -> (f64, f64, f64) {
    if xs.len() < 2 {
        return (1.0, 0.0, 1.0);
    }
    let n = xs.len() as f64;
    let mut sxx = 0.0;
    let mut sxy = 0.0;
    let mut sx = 0.0;
    let mut sy = 0.0;
    for i in 0..xs.len() {
        sxx += xs[i] * xs[i];
        sxy += xs[i] * ys[i];
        sx += xs[i];
        sy += ys[i];
    }
    let denom = n * sxx - sx * sx;
    let slope = if denom != 0.0 {
        (n * sxy - sx * sy) / denom
    } else {
        1.0
    };
    let intercept = (sy - slope * sx) / n;
    let y_bar = sy / n;
    let mut ss_res = 0.0;
    let mut ss_tot = 0.0;
    for i in 0..xs.len() {
        let pred = slope * xs[i] + intercept;
        ss_res += (ys[i] - pred).powi(2);
        ss_tot += (ys[i] - y_bar).powi(2);
    }
    let r2 = if ss_tot > 0.0 { 1.0 - ss_res / ss_tot } else { 1.0 };
    (slope, intercept, r2)
}

pub fn build_dual_clock(config: DualClockConfig<'_>) -> DualClockResult {
    let hamiltonian = config.hamiltonian;
    let sites = hamiltonian.n;
    let embed_dim = config.embed_dim;
    let order = config.order;

    let trajectory = evolve_trajectory(
        hamiltonian,
        &config.initial,
        config.reference.as_ref(),
        config.dt,
        config.steps,
        order,
    );

    let base_z: Vec<f64> = (0..sites)
        .map(|q| expectation_z(&config.initial, q))
        .collect();
    let energy0 = hamiltonian.expectation(&config.initial);

    let uniform_indices: Vec<usize> = (0..trajectory.len()).collect();
    let uniform_times: Vec<f64> = trajectory.iter().map(|p| p.t).collect();
    let uniform_slice_list = slices_from_indices(
        hamiltonian,
        &trajectory,
        &uniform_indices,
        &uniform_times,
        &config.initial,
        embed_dim,
    );
    let mut energy_drift = 0.0_f64;
    for s in &uniform_slice_list {
        energy_drift = energy_drift.max((s.energy - energy0).abs());
    }

    let (phys_indices, _) =
        physical_clock_indices(&trajectory, config.clock_site, &base_z, config.physical_slices);
    let physical_times: Vec<f64> = (0..phys_indices.len()).map(|k| k as f64).collect();
    let physical_slice_list = slices_from_indices(
        hamiltonian,
        &trajectory,
        &phys_indices,
        &physical_times,
        &config.initial,
        embed_dim,
    );
    for s in &physical_slice_list {
        energy_drift = energy_drift.max((s.energy - energy0).abs());
    }

    let time_map: Vec<TimeMapPoint> = phys_indices
        .iter()
        .enumerate()
        .map(|(k_physical, &uniform_idx)| TimeMapPoint {
            tau_uniform: uniform_idx as f64,
            tau_physical: k_physical as f64,
        })
        .collect();

    let xs: Vec<f64> = time_map.iter().map(|p| p.tau_physical).collect();
    let ys: Vec<f64> = time_map.iter().map(|p| p.tau_uniform).collect();
    let (sync_slope, sync_intercept, sync_r2) = fit_affine(&xs, &ys);

    DualClockResult {
        sites,
        uniform: ClockHistory {
            label: "Uniform clock (fixed Δt)".to_string(),
            clock_site: None,
            result: SpacetimeResult {
                sites,
                slices: uniform_slice_list,
                energy_drift,
                light_cone: None,
                elapsed_ms: 0.0,
            },
        },
        physical: ClockHistory {
            label: format!("Physical clock (site {})", config.clock_site),
            clock_site: Some(config.clock_site),
            result: SpacetimeResult {
                sites,
                slices: physical_slice_list,
                energy_drift,
                light_cone: None,
                elapsed_ms: 0.0,
            },
        },
        time_map,
        sync_r2,
        sync_slope,
        sync_intercept,
        elapsed_ms: 0.0,
    }
}
